dd-sdk-android/src/main/kotlin/com/datadog/android/DatadogInterceptor.kt

```diff
@@ -51,14 +57,127 @@ import okhttp3.OkHttpClient
+    localTracerFactory: () -> Tracer
+) : TracingInterceptor(tracedHosts, tracedRequestListener, localTracerFactory) {
+
+    /**
+     * Creates a [TracingInterceptor] to automatically create a trace around OkHttp [Request]s, and
+     * track RUM Resources.
+     *
+     * @param tracedHosts a list of all the hosts that you want to be automatically tracked
+     * by our APM [TracingInterceptor]. If no host provided the interceptor won't trace
+     * any OkHttpRequest, nor propagate tracing information to the backend.
+     * Please note that the host constraint will only be applied on the TracingInterceptor and we will
+     * continue to dispatch RUM Resource events for each request without applying any host filtering.
+     * @param tracedRequestListener which listens on the intercepted [okhttp3.Request] and offers
+     * the possibility to modify the created [io.opentracing.Span].
+     */
+    @JvmOverloads
+    constructor(
+        tracedHosts: List<String>,
+        tracedRequestListener: TracedRequestListener = NoOpTracedRequestListener()
+    ) : this(tracedHosts, tracedRequestListener, { AndroidTracer.Builder().build() })
+
+    // region Interceptor
+
+    /** @inheritdoc */
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+        val url = request.url().toString()
+        val method = request.method()
+        val requestId = identifyRequest(request)
+        println("stopResource $requestId $method $url")
+
+        GlobalRum.get().startResource(requestId, method, url)
+
+        return super.intercept(chain)
+    }
+
+    // endregion
+
+    // region TracingInterceptor
+
+    /** @inheritdoc */
+    override fun onRequestIntercepted(
+        request: Request,
+        span: Span,
+        response: Response?,
+        throwable: Throwable?
+    ) {
+        super.onRequestIntercepted(request, span, response, throwable)
+
+        if (throwable != null) {
+            handleThrowable(request, throwable)
+        } else {
+            handleResponse(request, response, span)
+        }
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun handleResponse(
+        request: Request,
+        response: Response?,
+        span: Span
+    ) {
+        val requestId = identifyRequest(request)
+        val statusCode = response?.code()
+        val method = request.method()
+        val mimeType = response?.header(HEADER_CT)
+        val kind = when {
+            method in xhrMethods -> RumResourceKind.XHR
+            mimeType == null -> RumResourceKind.UNKNOWN
+            else -> RumResourceKind.fromMimeType(mimeType)
+        }
+        println("stopResource $requestId $statusCode ${getBodyLength(response)} $kind")
+        GlobalRum.get().stopResource(
+            requestId,
+            statusCode,
+            getBodyLength(response),
+            kind,
+            mapOf(
+                RumAttributes.TRACE_ID to span.context().toTraceId()
```

Comment:
don't we inject `traceID` to `request.headers` by `TracingInterceptor`?
so with `CombinedInterceptor`, we could extract `traceID` from `request` in `RUMInterceptor`, am i wrong?

why did we remove `CombinedInterceptor` concept? because i was thinking of implementing it in iOS

Reply:
Well the problem is that the CombinedInterceptor would be using two RequestInterceptor which would know nothing about one another. It was indeed a really nice concept, but based on the idea that RUM and APM would be entirely decoupled. The rules on that have changed, and they will be much more tightly linked, so to ease things, I decided to use a more coupled approach in the code too.